### PR TITLE
COVID Custom case close rules

### PR DESCRIPTION
**Pull request by proteusvacuum**
https://docs.google.com/document/d/1LLrD_jgtdP6VmoxvjqGEEPTvsDGv9WyhNzQcHrDkD8c/edit?ts=5f0ce3e9#

##### SUMMARY

Adds a custom filter and custom action for use in the auto case updater. For COVID work. 
Requires Case Search Index to be initialized for the projects that will use this. 

Dependent on https://github.com/dimagi/commcare-cloud/pull/4017

### CODE REVIEW

**Review by orangejenny — COMMENTED** [review]
Just minor feedback, nothing blocking.

The filter looks like it could be useful for other projects - I'm not saying it should be in `corehq` right now, just musing.

**Review by orangejenny — APPROVED** [review]

File: custom/covid/rules/custom_criteria.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    username = checkin_case.get_case_property("username")
+    query = (
+        CaseSearchES()
+        .domain(checkin_case.domain)
+        .case_type(USERCASE_TYPE)
+        .case_property_query("username", username)
+    )
+    results = query.run().hits
+    if not results:
+        raise Exception
```
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply custom/covid/rules/custom_criteria.py]: Oops, I meant to handle this differently. Updated.

File: custom/covid/rules/custom_criteria.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from corehq.apps.es.case_search import CaseSearchES, flatten_result
+from casexml.apps.case.models import CommCareCase
+from corehq.apps.app_manager.const import USERCASE_TYPE
+
+
+def associated_user_cases_closed(case, now):
```
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply custom/covid/rules/custom_criteria.py]: :+1: Updated

File: custom/covid/tests/test_custom_rules.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+class DeactivatedMobileWorkersTest(BaseCaseRuleTest):
+    def setUp(self):
+        super().setUp()
+        delete_all_users()
+
+        self.domain = "covid-deactivated-mobile-workers"
+        self.domain_obj = create_domain(self.domain)
+        enable_usercase(self.domain)
+
+        with trap_extra_setup(ConnectionError):
```
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply custom/covid/tests/test_custom_rules.py:38]: I don't think this matters anymore, but this test would get skipped if ES isn't set up in that particular test environment. A lot of our other ES tests do it, so I just maintained that convention.

**Pull request merged**